validator/project_validator.go

```diff
@@ -449,6 +441,113 @@ func validateProjectConfigAliases(pc *model.ProjectConfig) ValidationErrors {
+	for _, a := range aliases {
+		aliasMap[a.ID.Hex()] = a
+	}
+	aliasNeedsVariant, aliasNeedsTask, err := getAliasCoverage(p, aliasMap)
+	if err != nil {
+		return ValidationErrors{
+			{
+				Message: "error checking alias coverage, continuing without validation",
+				Level:   Warning,
+			},
+		}
+	}
+	return constructAliasWarnings(aliasMap, aliasNeedsVariant, aliasNeedsTask)
+}
+
+// constructAliasWarnings returns validation errors given a map of aliases, and whether they need variants/tasks to match.
+func constructAliasWarnings(aliasMap map[string]model.ProjectAlias, aliasNeedsVariant, aliasNeedsTask map[string]bool) ValidationErrors {
+	res := ValidationErrors{}
+	for aliasID, a := range aliasMap {
+		needsVariant := aliasNeedsVariant[aliasID]
+		needsTask := aliasNeedsTask[aliasID]
+		if !needsVariant && !needsTask {
+			continue
+		}
+
+		msgComponents := []string{"Commit queue alias"}
+		if len(a.VariantTags) > 0 {
+			msgComponents = append(msgComponents, fmt.Sprintf("matching variant tags '%v'", a.VariantTags))
+		} else {
+			msgComponents = append(msgComponents, fmt.Sprintf("matching variant regexp '%s'", a.Variant))
+		}
+		if needsVariant {
+			msgComponents = append(msgComponents, "has no matching variants")
+		} else {
+			// This is only relevant information if the alias matches the variant but not the task.
+			if len(a.TaskTags) > 0 {
+				msgComponents = append(msgComponents, fmt.Sprintf("and matching task tags '%v'", a.TaskTags))
+			} else {
+				msgComponents = append(msgComponents, fmt.Sprintf("and matching task regexp '%s'", a.Task))
+			}
+			msgComponents = append(msgComponents, "has no matching tasks")
+		}
+		res = append(res, ValidationError{
+			Message: strings.Join(msgComponents, " "),
+			Level:   Warning,
+		})
+	}
+	return res
+}
+
+// getAliasCoverage returns a map of aliases that don't match variants and a map of aliases that don't match tasks.
+func getAliasCoverage(p *model.Project, aliasMap map[string]model.ProjectAlias) (map[string]bool, map[string]bool, error) {
+	type taskInfo struct {
+		name string
+		tags []string
+	}
+	aliasNeedsVariant := map[string]bool{}
+	aliasNeedsTask := map[string]bool{}
+	bvtCache := map[string]taskInfo{}
+	for a := range aliasMap {
+		aliasNeedsVariant[a] = true
+		aliasNeedsTask[a] = true
+	}
+	for _, bv := range p.BuildVariants {
+		for aliasID, alias := range aliasMap {
+			if !aliasNeedsVariant[aliasID] && !aliasNeedsTask[aliasID] { // Have already found both variants and tasks.
+				continue
+			}
+			// If we still need a task to match the variant, still check if the alias matches, so we know if checking tasks is needed.
+			matchesThisVariant, err := alias.HasMatchingVariant(bv.Name, bv.Tags)
+			if err != nil {
+				return nil, nil, err
+			}
+			if !matchesThisVariant { // If the variant doesn't match, then there's no reason to keep checking tasks.
+				continue
+			}
+			aliasNeedsVariant[aliasID] = false
+			for _, t := range bv.Tasks {
```

Comment:
Can we refactor this third loop to another helper function that takes in the `bvtCache`, `bv`, and `alias` and returning a bool/error?